File: model/blocksparse.py
------------------------------------------------------------------------
--[[ WindowSparse ]]--
-- A 3 layer model of Distributed Conditional Computation
------------------------------------------------------------------------
local BlockSparse, parent = torch.class("dp.BlockSparse", "dp.Layer")
BlockSparse.isBlockSparse = true

function BlockSparse:__init(config)
   assert(type(config) == 'table', "Constructor requires key-value arguments")
   local args, input_size, hidden_size, gater_size, output_size, 
      lr, sparse_init, typename, sparsityFactor, threshold_lr, alpha_range, std
      = xlua.unpack(
      {config},
      'BlockSparse', 
      'Deep mixture of experts model. It is three parametrized '..
      'layers of gated experts. There are two gaters, one for each '..
      'layer of hidden neurons.',
      {arg='input_size', type='number', req=true,
       help='Number of input neurons'},
      {arg='hidden_size', type='table', req=true,
       help='size of the hidden layers between nn.BlockSparses.'},
      {arg='gater_size', type='table', req=true,
       help='output size of the gaters'},
      {arg='output_size', type='number', req=true,
       help='output size of last layer'},
      {arg='lr', type='table', req=true,
       help='lr of gaussian blur target'},
      {arg='sparse_init', type='boolean', default=false,
       help='sparse initialization of weights. See Martens (2010), '..
       '"Deep learning via Hessian-free optimization"'}
      {arg='typename', type='string', default='BlockSparse', 
       help='identifies Model type in reports.'},
      {arg='sparsityFactor', type='number', default=0.1,
       help='sparsity for noisyrelu'},
      {arg='threshold_lr', type='number', default=0.1,
       help='learning rate to get the optimum threshold for a desired sparsity'},
      {arg='alpha_range', type='table', default={0.5, 1000, 0.01},
       help='{start_alpha, num_batches, endsall}'},
      {arg='std', type='number', req=true, default=0.1,
       help='std for gaussian noise'},
   )
   self._input_size = input_size
   self._output_size = output_size
   self._gater_size = gater_size
   self._hidden_size = hidden_size
   self._lr = lr
   
   
   require 'cunnx'
      
   --[[ First Layer : Input is dense, output is sparse ]]--
   -- Gater A
   local gaterA = nn.Sequential()
   gaterA:add(nn.Linear(self._input_size, self._gater_size[1]))
   local gateA = nn.NoisyReLU(opt.sparsityFactor, opt.threshold_lr, opt.alpha_range, opt.std)
   gaterA:add(gateA)
   gaterA:add(nn.SortFilter(opt.sparsityFactor))
   
   -- Mixture of experts A
   local concatA = nn.ConcatTable() -- outputs a table of tensors
   concatA:add(nn.Identity()) -- forwards input as is
   concatA:add(gaterA) -- forwards gated expert indices
   
   local mixtureA = nn.Sequential()
   mixtureA:add(concatA)
   expertsA = nn.BlockSparse(self._input_size, self._hidden_size[1], 1, self._gater_size[1])
   mixtureA:add(expertsA)
   mixtureA:add(nn.Tanh)
   

   --[[ Second Layer : Input and output are sparse ]]--
   -- Gater B : The input to the B gater is sparse so we use a nn.BlockSparse instead of a nn.Linear:
   local gaterB = nn.Sequential()
   gaterB:add(nn.BlockSparse(self._hidden_size[1]), self._gater_size[2], self._gater_size[1], 1))
   gateB = nn.NoisyReLU(opt.sparsityFactor, opt.threshold_lr, opt.alpha_range, opt.std)
   gaterB:add(gateB)
   gaterB:add(nn.SortFilter(opt.sparsityFactor))

   
   -- Mixture of experts B
   local concatB = nn.ConcatTable() -- outputs a table of tensors
   concatB:add(nn.Identity()) -- forwards the input to the output
   concatB:add(gaterB)

   local mixtureB = nn.Sequential()
   mixtureB:add(concatB)
   expertsB = nn.BlockSparse(self._hidden_size[1]), self._gater_size[2], self._gater_size[1], self._gater_size[2])
   mixtureB:add(expertsB)
   mixture:add(nn.Tanh)

   --[[ Third Layer : Input is sparse, output is dense ]]--
   -- Mixture of experts C
   local mixtureC = nn.Sequential()
   expertsC = nn.BlockSparse(self._hidden_size[2], self._output_size, self._gater_size[2], 1)
   mixtures:add(expertsC)
   mixtureC:add(nn.Tanh())

   --[[ Stack Mixtures ]]--
   -- Stack the 3 mixtures of experts layers. 
   local mlp = nn.Sequential()
   mlp:add(mixtureA)
   mlp:add(mixtureB)
   mlp:add(mixtureC)
   
   self._gaterA = gaterA
   self._gaterB = gaterB
   self._gateA = gateA
   self._gateB = gateB
   self._expertsA = expertsA
   self._expertsB = expertsB
   self._expertsC = expertsC
   self._mixtureA = mixtureA
   self._mixtureB = mixtureB
   self._mixtureC = mixtureC
   self._module = mlp
   
   config.typename = typename
   config.output = dp.DataView()
   config.input_view = 'bf'
   config.output_view = 'bf'
   config.tags = config.tags or {}
   config.tags['no-momentum'] = true
   config.sparse_init = sparse_init
   parent.__init(self, config)
end

-- requires targets be in carry
function BlockSparse:_forward(carry)
   self:outputAct(self._module:forward(self:inputAct()))
   return carry
end

function BlockSparse:_backward(carry)
   self._acc_scale = carry.scale or 1
   self._report.scale = self._acc_scale
   local input_act = self:inputAct()
   local output_grad = self:outputGrad()
   -- we don't accGradParameters as updateParameters will do so inplace
   output_grad = self._module:updateGradInput(input_act, output_grad)
   self:inputGrad(output_grad)
   return carry
end

function BlockSparse:paramModule()
   return self._module
end

function BlockSparse:_type(type)
   self._input_type = type
   self._output_type = type
   self._module:type(type)
   return self
end

function BlockSparse:reset()
   self._module:reset()
   if self._sparse_init then
      self._sparseReset(self._module.weight)
   end
end

function BlockSparse:zeroGradParameters()
   self._module:zeroGradParameters(true)
end

-- if after feedforward, returns active parameters 
-- else returns all parameters
function BlockSparse:parameters()
   return self._module:parameters(true)
end

function BlockSparse:sharedClone()
   local clone = torch.protoClone(self, {
      input_size=self._input_size, hierarchy={[1]=torch.IntTensor{1,2}},
      root_id=1, sparse_init=self._sparse_init,
      dropout=self._dropout and self._dropout:clone(),
      typename=self._typename, 
      input_type=self._input_type, output_type=self._output_type,
      module_type=self._module_type, mvstate=self.mvstate
   })
   clone._target_type = self._target_type
   clone._module = self._module:sharedClone()
   return clone
end

function BlockSparse:updateParameters(lr)
   -- we update parameters inplace (much faster)
   -- so don't use this with momentum (disabled by default)
   self._module:accUpdateGradParameters(self:inputAct(), self.outputGrad(), self._acc_scale * lr)
   return
end

File: model/blocksparse.lua
------------------------------------------------------------------------
--[[ BlockSparse ]]--
-- A n-layer model for Distributed Conditional Computation
------------------------------------------------------------------------
local BlockSparse, parent = torch.class("dp.BlockSparse", "dp.Layer")
BlockSparse.isBlockSparse = true

function BlockSparse:__init(config)
   assert(type(config) == 'table', "Constructor requires key-value arguments")
   local args, input_size, n_block, hidden_size, window_size, gater_size, 
      output_size, noise_std, gater_act, expert_act, gater_style, 
      expert_scale, gater_scale, interleave, threshold_lr, alpha_range, 
      sparse_init, typename = xlua.unpack(
      {config},
      'BlockSparse', 
      'Deep mixture of experts model. It is n parametrized '..
      'layers of gated experts. There are n-1 gaters, one for each '..
      'layer of hidden neurons between nn.BlockSparses.',
      {arg='input_size', type='number', req=true,
       help='Number of input neurons'},
      {arg='n_block', type='table', req=true,
       help='number blocks in hidden layers between nn.BlockSparses.'},
      {arg='hidden_size', type='table', req=true,
       help='number of neurons per block in hidden layers between nn.BlockSparses.'},
      {arg='window_size', type='table', req=true,
       help='number of blocks used per example in each layer.'},
      {arg='gater_size', type='table', req=true,
       help='number of neurons in gater hidden layers'},
      {arg='output_size', type='number', req=true,
       help='output size of last layer'},
      {arg='noise_std', type='number', req=true,
       help='std deviation of gaussian noise used for NoisyReLU'},
      {arg='gater_act', type='nn.Module', default='',
       help='gater hidden activation. Defaults to nn.Tanh()'},
      {arg='expert_act', type='table', default='',
       help='expert activation. Defaults. to nn.Tanh()'},
      {arg='gater_style', type='string', default='NoisyReLU',
       help='comma-separated sequence of Modules to use for gating'},
      {arg='expert_scale', type='number', default=1,
       help='scales the learningRate for the experts'},
      {arg='gater_scale', type='number', default=1,
       help='scales the learningRate for the gater'},
      {arg='interleave', type='boolean', default=false,
       help='when true, alternate between training gater and experts every epoch'},
      {arg='threshold_lr', type='number', default=0.1,
       help='learning rate to get the optimum threshold for a desired sparsity'},
      {arg='alpha_range', type='table', default='',
       help='{start_alpha, num_batches, endsall}'},
      {arg='sparse_init', type='boolean', default=false,
       help='sparse initialization of weights. See Martens (2010), '..
       '"Deep learning via Hessian-free optimization"'},
      {arg='typename', type='string', default='BlockSparse', 
       help='identifies Model type in reports.'}
   )
   self._input_size = input_size
   self._output_size = output_size
   self._gater_size = gater_size
   self._n_block = n_block
   self._hidden_size = hidden_size
   self._window_size = window_size
   self._interleave = interleave
   self._expert_scale = expert_scale
   self._gater_scale = gater_scale
   self._expert_phase = true
   alpha_range = (alpha_range == '') and {0.5, 1000, 0.01} or alpha_range
   gater_act = (gater_act == '') and nn.Tanh() or gater_act
   expert_act = (expert_act == '') and nn.Tanh() or expert_act
   
   -- experts
   self._experts = {}
   local para = nn.ParallelTable()
   para:add(expert_act:clone())
   para:add(nn.Identity())
   
   local expert = nn.Sequential()
   expert:add(nn.BlockSparse(1, self._input_size, self._n_block[1], self._hidden_size[1], config.acc_update))
   expert:add(para)
   table.insert(self._experts, expert)
   
   for i=1,#self._n_block - 1 do
      expert = nn.Sequential()
      expert:add(nn.BlockSparse(self._n_block[i], self._hidden_size[i], self._n_block[i+1], self._hidden_size[i+1], config.acc_update))
      expert:add(para:clone())
      table.insert(self._experts, expert)
   end
   
   local dept = #self._n_block
   expert = nn.Sequential()
   expert:add(nn.BlockSparse(self._n_block[dept], self._hidden_size[dept], 1, output_size, config.acc_update))
   expert:add(expert_act:clone())
   table.insert(self._experts, expert)
   
   -- gaters
   self._gates = {}
   
   self._gater = nn.Sequential()
   local inputSize = self._input_size
   for i, gaterSize in ipairs(self._gater_size) do
      self._gater:add(nn.Linear(inputSize, gaterSize))
      self._gater:add(gater_act:clone())
      inputSize = gaterSize
   end
   
   local concat = nn.ConcatTable()
   self._relus = {}
   self._balances = {}
   for i=1,#self._window_size do
      local gate = nn.Sequential()
      gate:add(nn.Linear(inputSize, self._n_block[i]))
      for i,gater_str in ipairs(_.split(gater_style, '[,]')) do
         if gater_str == 'NoisyReLU' then
            local relu = nn.NoisyReLU(self._window_size[i]/self._n_block[i], threshold_lr, alpha_range, noise_std[i])
            gate:add(relu)
            table.insert(self._relus, relu)
         elseif gater_str == 'Balance' then
            local balance = nn.Balance(10)
            gate:add(balance)
            table.insert(self._balances, balance)
         elseif gater_str == 'SoftMax' then
            gater:add(nn.SoftMax())
         else
            error("unknown gater_style:"..gater_str, 2)
         end
      end
      gate:add(nn.LazyKBest(self._window_size[i]))
      concat:add(gate)
      table.insert(self._gates, gate)
   end
   self._gater:add(concat)
   
   -- mixture
   gater_scale = interleave and 0 or gater_scale
   self._bm = nn.BlockMixture(self._experts, self._gater, expert_scale, gater_scale)
   self._module = self._bm
   
   config.typename = typename
   config.output = dp.DataView()
   config.input_view = 'bf'
   config.output_view = 'bf'
   config.tags = config.tags or {}
   config.sparse_init = sparse_init
   parent.__init(self, config)
   -- only works with cuda
   self:type('torch.CudaTensor')
end

function BlockSparse:doneEpoch(report, ...)
   self:zeroStatistics()
   if self._interleave then
      self._expert_phase = not self._expert_phase
      self._bm.expertScale = self._expert_phase and self._expert_scale or 0
      self._bm.gaterScale = (not self._expert_phase) and self._gater_scale or 0
   end
end

function BlockSparse:report()
   if not self._next_report then
      self._next_report = true
      return
   end
   if self._expert_phase then
      print"Expert Phase"
   else
      print"Gater Phase"
   end
   local nVal = 5
   for i, relu in ipairs(self._relus) do
      local vals = relu.sparsity:select(1,1):float():sort()
      print(i, 
         table.tostring(vals:narrow(1,1,nVal):clone():storage():totable()), 
         table.tostring(vals:narrow(1,vals:size(1)-nVal+1,nVal):clone():storage():totable())
      )
   end
end
